src/cli/self_update/shell.rs

```diff
@@ -0,0 +1,192 @@
+//! a trend has emerged to place PATH updates in other run-commands (rc) files,
+//! leaving Rustup with few assumptions to build on for fulfilling its promise
+//! to set up PATH appropriately.
+//!
+//! Rustup addresses this by:
+//! 1) using a shell script that updates PATH if the path is not in PATH
+//! 2) sourcing this script in any known and appropriate rc file
+
+use super::canonical_cargo_home;
+use super::*;
+use crate::process;
+use std::path::PathBuf;
+
+pub type Shell = Box<dyn UnixShell>;
+
+#[derive(Debug, PartialEq)]
+pub struct ShellScript {
+    content: &'static str,
+    name: &'static str,
+}
+
+impl ShellScript {
+    pub fn write(&self) -> Result<()> {
+        let cargo_bin = format!("{}/bin", canonical_cargo_home()?);
+        let env_name = utils::cargo_home()?.join(self.name);
+        let env_file = self.content.replace("{cargo_bin}", &cargo_bin);
+        utils::write_file(self.name, &env_name, &env_file)?;
+        Ok(())
+    }
+}
+
+#[allow(dead_code)] // For some reason.
+const POSIX_ENV: &str = include_str!("env");
+
+macro_rules! support_shells {
+    ( $($shell:ident,)* ) => {
+        fn enumerate_shells() -> Vec<Shell> {
+            vec![$( Box::new($shell), )*]
+        }
+    }
+}
+
+// TODO: Tcsh (BSD)
+// TODO?: Make a decision on Ion Shell, Power Shell, Nushell
+// Cross-platform non-POSIX shells have not been assessed for integration yet
+support_shells! {
+    Posix,
+    Bash,
+    Zsh,
+}
+
+pub fn get_available_shells() -> impl Iterator<Item = Shell> {
+    enumerate_shells().into_iter().filter(|sh| sh.does_exist())
+}
+
+pub trait UnixShell {
+    // Detects if a shell "exists". Users have multiple shells, so an "eager"
+    // heuristic should be used, assuming shells exist if any traces do.
+    fn does_exist(&self) -> bool;
+
+    // Gives all rcfiles of a given shell that rustup is concerned with.
+    // Used primarily in checking rcfiles for cleanup.
+    fn rcfiles(&self) -> Vec<PathBuf>;
+
+    // Gives rcs that should be written to.
+    fn update_rcs(&self) -> Vec<PathBuf>;
+
+    // Writes the relevant env file.
+    fn env_script(&self) -> ShellScript {
+        ShellScript {
+            name: "env",
+            content: POSIX_ENV,
+        }
+    }
+
+    fn source_string(&self) -> Result<String> {
+        Ok(format!(r#"source "{}/env""#, canonical_cargo_home()?))
+    }
+}
+
+struct Posix;
+impl UnixShell for Posix {
+    fn does_exist(&self) -> bool {
+        true
+    }
+
+    fn rcfiles(&self) -> Vec<PathBuf> {
+        match utils::home_dir() {
+            Some(dir) => vec![dir.join(".profile")],
+            _ => vec![],
+        }
+    }
+
+    fn update_rcs(&self) -> Vec<PathBuf> {
+        // Write to .profile even if it doesn't exist.
+        self.rcfiles()
+    }
+}
+
+struct Bash;
+
+// Bash is the source of many complications because it loads ONLY 1 rc,
+// either a login profile or .bashrc, not both.
```

Comment:
It is true that it does that, but every bash_profile I've ever seen has 
```

if [ -f ~/.bashrc ]; then
    source ~/.bashrc
fi

```

in it, so I'm really not sure that this comment is more than us being judgemental, in a way that ill suits us.

Reply:
Noted. Fiiixed?